src/sources/postgresql_metrics.rs

```diff
@@ -0,0 +1,867 @@
+    if keepalives_idle != 2 * 60 * 60 {
+        params.push(("keepalives_idle", keepalives_idle.to_string()));
+    }
+
+    // target_session_attrs, ignore default value (TargetSessionAttrs::Any)
+    match config.get_target_session_attrs() {
+        TargetSessionAttrs::Any => {} // default, ignore
+        TargetSessionAttrs::ReadWrite => {
+            params.push(("target_session_attrs", "read-write".to_owned()))
+        }
+        // non_exhaustive enum
+        _ => {
+            warn!(r#"Unknown variant of "TargetSessionAttrs.""#);
+        }
+    }
+
+    // channel_binding, ignore default value (ChannelBinding::Prefer)
+    match config.get_channel_binding() {
+        ChannelBinding::Disable => params.push(("channel_binding", "disable".to_owned())),
+        ChannelBinding::Prefer => {} // default, ignore
+        ChannelBinding::Require => params.push(("channel_binding", "require".to_owned())),
+        // non_exhaustive enum
+        _ => {
+            warn!(r#"Unknown variant of "ChannelBinding"."#);
+        }
+    }
+
+    format!(
+        "postgresql:///{}?{}",
+        config.get_dbname().unwrap_or(""),
+        params
+            .into_iter()
+            .map(|(k, v)| format!(
+                "{}={}",
+                percent_encoding(k.as_bytes()),
+                percent_encoding(v.as_bytes())
+            ))
+            .collect::<Vec<String>>()
+            .join("&")
+    )
+}
+
+fn percent_encoding(input: &'_ [u8]) -> String {
+    percent_encoding::percent_encode(input, percent_encoding::NON_ALPHANUMERIC).to_string()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn generate_config() {
+        crate::test_util::test_generate_config::<PostgresqlMetricsConfig>();
+    }
+}
+
+#[cfg(all(test, feature = "postgresql_metrics-integration-tests"))]
+mod integration_tests {
+    use super::*;
+    use crate::{test_util::trace_init, Pipeline};
+
+    async fn test_postgresql_metrics(endpoint: String, tls: Option<PostgresqlMetricsTlsConfig>) {
+        trace_init();
+
+        let (sender, mut recv) = Pipeline::new_test();
+
+        tokio::spawn(async move {
+            PostgresqlMetricsConfig {
+                endpoints: vec![endpoint],
+                tls,
+                ..Default::default()
+            }
+            .build(
+                "default",
+                &GlobalOptions::default(),
+                ShutdownSignal::noop(),
+                sender,
+            )
+            .await
+            .unwrap()
+            .await
+            .unwrap()
+        });
+
+        let event = time::timeout(time::Duration::from_secs(3), recv.next())
+            .await
+            .expect("fetch metrics timeout")
+            .expect("failed to get metrics from a stream");
+        let mut events = vec![event];
+        loop {
+            match time::timeout(time::Duration::from_millis(10), recv.next()).await {
+                Ok(Some(event)) => events.push(event),
```

Comment:
Sorry, just seeing taking a closer look at the tests now, could we expand this test to verify some additional things about the metrics?

Maybe:

* Making sure `up` returns 1
* Making sure that some known metric is always returned
* Making sure they are all tagged correctly with any tags that are applied to all
* Testing the negative case to make sure `up` is reported with 0
* Making sure the namespace is applied correctly to all metrics